File: tagalong/020-methods/methods.go
package main

import "fmt"

type rectangle struct {
	width  float64
	height float64
}

func (r rectangle) area() float64 {
	return r.width * r.height
}

func (r rectangle) perim() float64 {
	return 2*r.width + 2*r.height
}

func (r *rectangle) scale(scale float64) {
	r.height *= scale
	r.width *= scale
}

func main() {
	r := rectangle{width: 12.7, height: 10}
	fmt.Println("area [mm²]:", r.area())
	fmt.Println("perimeter [mm]:", r.perim())

	// Convert to inches.
	r.scale(1. / 25.4)
	fmt.Println("area[inches²]:", r.area())
	fmt.Println("perim[inches]:", r.perim())
}

File: tagalong/020-methods/methods.py
class Rectangle:
    def __init__(self, width: float, height: float) -> None:
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def perim(self) -> float:
        return 2 * (self.width + self.height)

    def scale(self, factor: float) -> None:
        self.width *= factor
        self.height *= factor


if __name__ == "__main__":
    r = Rectangle(12.7, 10.0)
    print("area [mm²]:", r.area())
    print("perimeter [mm]:", r.perim())

    r.scale(1 / 25.4)
    print("area [in²]:", r.area())
    print("perimeter [in]:", r.perim())
